File: java/dsa/algorithms/dp/2d/coin-change-i-i.java
class Solution {
    public int change(int amount, int[] coins) {
        int[] dp = new int[amount + 1]; // @viz: init
        dp[0] = 1; // @viz: base
        
        for (int coin : coins) { // @viz: coin_loop
            for (int i = coin; i <= amount; i++) { // @viz: i_check
                dp[i] += dp[i - coin]; // @viz: i_update
            }
        }
        return dp[amount]; // @viz: result
    }
}

File: javascript/dsa/algorithms/dp/2d/coin-change-i-i.js
function change(amount, coins) {
    // 1️⃣ Initialize DP table
    const dp = Array(amount + 1).fill(0); // @viz: init
    dp[0] = 1; // @viz: base
    
    // 2️⃣ Outer loop: Coins (guarantees combinations, not permutations)
    for (let coin of coins) { // @viz: coin_loop
        for (let i = coin; i <= amount; i++) { // @viz: i_check
            dp[i] += dp[i - coin]; // @viz: i_update
        }
    }
    
    return dp[amount]; // @viz: result
}